plugins/module_utils/network/sonic/facts/snmp/snmp.py

```diff
@@ -0,0 +1,406 @@
+            snmp_dict.update({'location': self.get_snmp_location(snmp_list)})
+            snmp_dict.update({'enable_trap': list(self.get_snmp_enable_trap(snmp_list))})
+            snmp_dict.update({'group': self.get_snmp_group(snmp_list)})
+            host = self.get_snmp_hosts_targets(snmp_list)
+            snmp_dict.update({'host': host})
+
+        return snmp_dict
+
+    def get_snmp_agentaddress(self, snmp_list):
+        """
+        Get snmp agent address from the snmp list
+        """
+        agentaddress_dict = {}
+        agentaddress_list = []
+
+        if not snmp_list.get('engine') or not snmp_list.get('engine').get('listen'):
+            return agentaddress_list
+        agentaddress_config = snmp_list['engine']['listen']
+
+        for agentaddress in agentaddress_config:
+            agentaddress_dict['interface'] = agentaddress.get('udp').get("ietf-snmp-ext:interface")
+            agentaddress_dict['ip'] = agentaddress.get('udp').get("ip")
+            agentaddress_dict['port'] = agentaddress.get('udp').get("port")
+            autogen_agentaddress.append({'agent-entry-name': agentaddress.get('name'), 'address': {'ip': agentaddress_dict['ip']}})
+
+            agentaddress_list.append(agentaddress_dict)
+        return agentaddress_list
+
+    def get_snmp_community(self, snmp_list):
+        """
+        Get snmp community from the snmp list
+        """
+        community_list = []
+
+        if not snmp_list.get('community'):
+            return community_list
+        community_config = snmp_list['community']
+
+        for community in community_config:
+            if community is None:
+                return community_list
+            community_list.append({"name": community.get("index"), "group": community.get("security-name")})
+
+        return community_list
+
+    def get_snmp_engine(self, snmp_list):
+        """
+        Get snmp engine from the snmp list
+        """
+        engine = ''
+
+        if not snmp_list.get('engine'):
+            return engine
+
+        engine_config = snmp_list['engine'].get('engine-id')
+
+        return engine_config
+
+    def get_snmp_users(self, snmp_list):
+        """
+        Get snmp users from the snmp list
+        """
+        user_list = []
+        group_config = []
+
+        if not snmp_list.get('usm') or not snmp_list.get('usm').get('local') or not snmp_list.get('usm').get('local').get('user'):
+            return user_list
+
+        if snmp_list.get('vacm') and snmp_list.get('vacm').get('group'):
+            group_config = snmp_list['vacm']['group']
+
+        user_config = snmp_list['usm']['local'].get('user')
+
+        for user in user_config:
+            user_dict = {}
+            auth_type = list(user.get('auth').keys())[0]
+            auth_key = user.get("auth").get(auth_type).get('key')
+
+            priv_type = list(user.get('priv').keys())[0]
+            priv_key = user.get("priv").get(priv_type).get('key')
```

Comment:
Please handle optional fields 'auth' and 'priv'.
```suggestion
            if user.get('auth'):
                auth_type = list(user['auth'].keys())[0]
                user_dict['auth'] = {
                    'auth_type': auth_type,
                    'key': user['auth'][auth_type].get('key')
                }
            if user.get('priv'):
                priv_type = list(user['priv'].keys())[0]
                user_dict['priv'] = {
                    'priv_type': priv_type,
                    'key': user['priv'][priv_type].get('key')
                }
```